File: day7/day07.R
#!/usr/bin/Rscript
number_vec <- readLines("data/data_07.txt", n = 1) |>
    strsplit(",") |>
    unlist() |>
    as.numeric()
minmax <- min(number_vec):max(number_vec)
# Q1
sapply(minmax, FUN = function(x) sum(abs(number_vec - x))) |> min()
# Q2
q2_fun <- function(x) {
    dist <- abs(number_vec - x)
    sum(((1 + dist) * dist) / 2)
}
sapply(minmax, FUN = q2_fun) |> min()

File: day7/day07.py
#!/usr/bin/env python3
# https://adventofcode.com/2021/day/7
import numpy as np

with open("data/data_07.txt", mode="r") as raw:
    arr = np.array(list(map(int, raw.readline().split(","))))


def Q1_get_minimun(arr: np.ndarray):
    uniqarr = np.unique(arr)
    return np.abs(arr - uniqarr[:, np.newaxis]).sum(axis=1).min()


def Q2_get_minimun(arr: np.ndarray):
    # this is far more fast than np.arange(0, x + 1).sum()
    # Math is matter!
    cum_sum = lambda x: ((1 + x) * x) / 2
    vec_cum_sum = np.vectorize(cum_sum)
    uniqarr = np.arange(arr.min(), arr.max())
    dist = np.abs(arr - uniqarr[:, np.newaxis])
    ans = vec_cum_sum(dist).sum(axis=1)
    return ans.min(), np.where(ans == ans.min())


print(Q1_get_minimun(arr))
print(arr)
print(Q2_get_minimun(arr))
